# Getting Started

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Matplotlib is building the font cache; this may take a moment.


In [6]:
# read csv into pandas dataframe
df = pd.read_csv("/Users/brisaniashley/Documents/School/AI4ALL Fall '24/AI4All_F24_Class16/big_startup_secsees_dataset.csv")

In [7]:
# get dimensions of df
df.shape

(66368, 14)

In [8]:
# get some sample records
df.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list funding_total_usd  \
0                                              Media          10000000   
1  Application Platforms|Real Time|Social Network...            700000   
2                                  Apps|Games|Mobile           3406878   
3                                        Curated Web           2000000   
4                                           Software                 -   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  
0               1         NaN       2015-01-05      2015-01-05  
1               2  2014-09-04       2014-03-01      2014-10-14  
2               1         NaN       2014-01-30      2014-01-30  
3               1  2007-01-01       2008-03-19      2008-03-19  
4               1  2010-01-01       2014-07-24      2014-07-24

In [32]:
df.sort_values(by=['founded_at'], ascending=True)

permalink  \
4776                               /organization/atipica   
59382               /organization/trinity-college-dublin   
24841                   /organization/harvard-university   
49957               /organization/scottish-newcastle-plc   
49105                   /organization/rutgers-university   
...                                                  ...   
66351            /organization/zynergy-projects-services   
66359                                /organization/zyrra   
66360                               /organization/zystor   
66363  /organization/zznode-science-and-technology-co...   
66367  /organization/İnovatiff-reklam-ve-tanıtım-hizm...   

                                             name  \
4776                                      Atipica   
59382                     Trinity College, Dublin   
24841                          Harvard University   
49957                    Scottish & Newcastle PLC   
49105                          Rutgers University   
...                                           ...   
66351                 Zynergy Projects & Services   
66359                                       Zyrra   
66360                                      Zystor   
66363               ZZNode Science and Technology   
66367  İnovatiff Reklam ve Tanıtım Hizmetleri Tic   

                       homepage_url                        category_list  \
4776          http://www.atipica.co                                  NaN   
59382            http://www.tcd.ie/  Colleges|Education|Public Relations   
24841            http://harvard.edu      Colleges|Education|Universities   
49957                           NaN                              Brewing   
49105        http://www.rutgers.edu                            Education   
...                             ...                                  ...   
66351  http://www.zynergygroup.net/                     Clean Technology   
66359          http://www.zyrra.com                           E-Commerce   
66360         http://www.zystor.com                        Biotechnology   
66363         http://www.zznode.com                  Enterprise Software   
66367          http://inovatiff.com   Consumer Goods|E-Commerce|Internet   

       funding_total_usd     status country_code state_code     region  \
4776             25000.0     closed          NaN        NaN        NaN   
59382        103198894.0  operating          NaN        NaN        NaN   
24841         90000000.0  operating          USA         MA     Boston   
49957          4200000.0  operating          GBR         U8  Edinburgh   
49105          1950000.0  operating          USA         NJ     Newark   
...                  ...        ...          ...        ...        ...   
66351                NaN  operating          IND         25    Chennai   
66359          1510500.0  operating          USA         MA     Boston   
66360          8500000.0   acquired          USA         WI  Milwaukee   
66363          1587301.0  operating          CHN         22    Beijing   
66367            14851.0  operating          NaN        NaN        NaN   

                city  funding_rounds  founded_at first_funding_at  \
4776             NaN               1  1015-01-30       2015-11-11   
59382            NaN               1  1592-01-01       2014-01-20   
24841      Cambridge               1  1636-09-08       2014-01-06   
49957      Edinburgh               1  1749-01-01       2003-10-10   
49105  New Brunswick               1  1766-01-01       2015-03-25   
...              ...             ...         ...              ...   
66351        Chennai               1         NaN       2015-08-19   
66359      Cambridge               4         NaN       2010-10-21   
66360      Milwaukee               1         NaN       2004-12-16   
66363        Beijing               1         NaN       2012-04-01   
66367            NaN               1         NaN       2013-10-01   

      last_funding_at  avg_funding_per_round  
4776       

In [31]:
df.sort_values(by=['first_funding_at'], ascending=True)

permalink              name  \
37176           /organization/mousera           Mousera   
22208         /organization/gamewheel         Gamewheel   
7409   /organization/blaze-bioscience  Blaze Bioscience   
61699             /organization/veveo             Veveo   
28378         /organization/intercomp         Intercomp   
...                               ...               ...   
47183       /organization/realmassive       RealMassive   
50717           /organization/sequent           Sequent   
52711           /organization/snapp-2     Snapp Builder   
53371            /organization/sontra            Sontra   
58592        /organization/topicmarks        Topicmarks   

                         homepage_url  \
37176          http://www.mousera.com   
22208             http://gamewheel.co   
7409   http://www.blazebioscience.com   
61699            http://www.veveo.net   
28378                             NaN   
...                               ...   
47183      http://www.realmassive.com   
50717          http://www.sequent.com   
52711            http://www.snapp.cc/   
53371        http://www.sontra.com.br   
58592           http://topicmarks.com   

                                           category_list  funding_total_usd  \
37176                                      Biotechnology         29940004.0   
22208  Brand Marketing|Game|Mobile Advertising|Social...                NaN   
7409   Biotechnology|Health Care|Health Care Informat...         32936998.0   
61699                             Mobile|Video Streaming         28000000.0   
28378                                                NaN           549000.0   
...                                                  ...                ...   
47183  Commercial Real Estate|Marketing Automation|Ma...          8000000.0   
50717                                           Software         12000000.0   
52711  Application Platforms|Development Platforms|Mo...           400000.0   
53371          Customer Service|Logistics|Transportation          4349586.0   
58592  Curated Web|Education|Knowledge Management|Nat...           150000.0   

          status country_code state_code         region           city  \
37176  operating          USA         CA    SF Bay Area      San Mateo   
22208  operating          DEU         16         Berlin         Berlin   
7409   operating          USA         WA        Seattle        Seattle   
61699   acquired          USA         MA         Boston        Andover   
28378     closed          NaN        NaN            NaN            NaN   
...          ...          ...        ...            ...            ...   
47183  operating          USA         TX         Austin         Austin   
50717  operating          USA         CA    SF Bay Area  Mountain View   
52711  operating          USA         NY  New York City       New York   
53371  operating          BRA          2      Sao Paulo      São Paulo   
58592   acquired          USA         CA    SF Bay Area  San Francisco   

       funding_rounds  founded_at first_funding_at last_funding_at  \
37176               3  2013-01-01       1014-03-01      2015-06-09   
22208               2         NaN       1201-11-23      2014-12-03   
7409                5  2010-01-01       1532-06-26      2013-12-03   
61699               3         NaN       1960-01-01      2007-05-01   
28378               1  1968-01-01       1971-01-01      1971-01-01   
...               ...         ...              ...             ...   
47183               2  2013-02-11              NaN      2015-05-14   
50717               3  2010-01-01              NaN      2013-07-08   
52711               3         NaN              NaN      2015-09-01   
53371               5  2013-01-01              NaN      2015-01-23   
58592               2  2009-01-01              NaN      2011-03-18   

       avg_funding_per_round  
37176           9.980001e+06  
22208                    NaN  
7409            6.587400e+06  
61699           9.33333

In [11]:
# get unique value of funding_rounds  
df['funding_rounds'].unique()

array([ 1,  2,  4,  3,  9,  5,  6,  7,  8, 10, 11, 12, 15, 13, 14, 17, 18,
       16, 19])

In [12]:
# get data types of columns
df.dtypes

permalink            object
name                 object
homepage_url         object
category_list        object
funding_total_usd    object
status               object
country_code         object
state_code           object
region               object
city                 object
funding_rounds        int64
founded_at           object
first_funding_at     object
last_funding_at      object
dtype: object

In [25]:
df['founded_at'] = pd.to_datetime(df['founded_at']) #bad date
df['first_funding_at'] = pd.to_datetime(df['first_funding_at']) #bad date

df['last_funding_at'] = pd.to_datetime(df['last_funding_at']) #successful

In [28]:
df.dtypes

permalink                        object
name                             object
homepage_url                     object
category_list                    object
funding_total_usd               float64
status                           object
country_code                     object
state_code                       object
region                           object
city                             object
funding_rounds                    int64
founded_at                       object
first_funding_at                 object
last_funding_at          datetime64[ns]
avg_funding_per_round           float64
dtype: object

In [26]:
# find distribution of fundtin_total_usd column
df['funding_total_usd'].describe()

count    5.358300e+04
mean     1.847860e+07
std      1.880133e+08
min      1.000000e+00
25%      3.356235e+05
50%      2.000000e+06
75%      1.000000e+07
max      3.007950e+10
Name: funding_total_usd, dtype: float64

In [14]:
df_by_countries = pd.DataFrame(df.groupby('country_code').size())
df_by_countries

0
country_code     
ALB             2
ARE            94
ARG           186
ARM             3
AUS           503
...           ...
VEN            14
VNM            41
ZAF            92
ZMB             2
ZWE             3

[137 rows x 1 columns]

In [15]:
df_by_countries.loc['USA']

0    37601
Name: USA, dtype: int64

# TO DO: 

- look into possible bias due to overrepresented data (in USA, for example)
- Delete: 37176	/organization/mousera	Mousera
- Delete: 22208	/organization/gamewheel	Gamewheel	http://gamewheel.co
- Delete: harvard & trinity college (dublin)
- Update: 7409	/organization/blaze-bioscience	Blaze Bioscience	http://www.blazebioscience.com, first round 2013-06-26 last round 2019-02-01, 6 funding rounds
- Updated: 4776	/organization/atipica	Atipica	http://www.atipica.co, founded_at 2015-01-30

# Data Cleaning

Missing values, either:
- drop rows
- drop columns
- fill in missing values with an appropriate value

In [16]:
# check columns for missing values in percentage
df.isnull().mean() * 100
# closest to zero is best, means 100% of data is filled.


permalink             0.000000
name                  0.001507
homepage_url          7.621143
category_list         4.743250
funding_total_usd     0.000000
status                0.000000
country_code         10.483968
state_code           12.878194
region               12.099204
city                 12.096191
funding_rounds        0.000000
founded_at           22.934245
first_funding_at      0.036162
last_funding_at       0.000000
dtype: float64

Columns with Missing Data In Order of Concern:

1. founded_at: 22.934245%
2. state_code: 12.878194%
3. region: 12.099204%
4. city: 12.096191%
5. country_code: 10.483968%
6. homepage_url: 7.621143%
7. category_list: 4.743250%
8. first_funding_at: 0.036162%
9. name: 0.001507%

In [17]:
# change data type of funding_total_usd to numerical
df['funding_total_usd'] = pd.to_numeric(df['funding_total_usd'], errors='coerce')
df.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list  funding_total_usd  \
0                                              Media         10000000.0   
1  Application Platforms|Real Time|Social Network...           700000.0   
2                                  Apps|Games|Mobile          3406878.0   
3                                        Curated Web          2000000.0   
4                                           Software                NaN   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  
0               1         NaN       2015-01-05      2015-01-05  
1               2  2014-09-04       2014-03-01      2014-10-14  
2               1         NaN       2014-01-30      2014-01-30  
3               1  2007-01-01       2008-03-19      2008-03-19  
4               1  2010-01-01       2014-07-24      2014-07-24

In [22]:
df.dtypes

permalink                 object
name                      object
homepage_url              object
category_list             object
funding_total_usd        float64
status                    object
country_code              object
state_code                object
region                    object
city                      object
funding_rounds             int64
founded_at                object
first_funding_at          object
last_funding_at           object
avg_funding_per_round    float64
dtype: object

# Feature Engineering

In [19]:
# calculate new column to find average funding per funding round
df['avg_funding_per_round'] = df['funding_total_usd'] / df['funding_rounds']
df.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list  funding_total_usd  \
0                                              Media         10000000.0   
1  Application Platforms|Real Time|Social Network...           700000.0   
2                                  Apps|Games|Mobile          3406878.0   
3                                        Curated Web          2000000.0   
4                                           Software                NaN   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  \
0               1         NaN       2015-01-05      2015-01-05   
1               2  2014-09-04       2014-03-01      2014-10-14   
2               1         NaN       2014-01-30      2014-01-30   
3               1  2007-01-01       2008-03-19      2008-03-19   
4               1  2010-01-01       2014-07-24      2014-07-24   

   avg_funding_per_round  
0             10000000.0  
1               350000.0  
2              3406878.0  
3              2000000.0  
4                    NaN

In [20]:
df.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list  funding_total_usd  \
0                                              Media         10000000.0   
1  Application Platforms|Real Time|Social Network...           700000.0   
2                                  Apps|Games|Mobile          3406878.0   
3                                        Curated Web          2000000.0   
4                                           Software                NaN   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  \
0               1         NaN       2015-01-05      2015-01-05   
1               2  2014-09-04       2014-03-01      2014-10-14   
2               1         NaN       2014-01-30      2014-01-30   
3               1  2007-01-01       2008-03-19      2008-03-19   
4               1  2010-01-01       2014-07-24      2014-07-24   

   avg_funding_per_round  
0             10000000.0  
1               350000.0  
2              3406878.0  
3              2000000.0  
4                    NaN

In [21]:
df.region.unique()

array(['Mumbai', 'DE - Other', nan, ..., 'Gardanne', 'Cancun', 'Muenchen'],
      dtype=object)